Fix active

Pakketeretet2 opened the pull request:
## Purpose

This adds a simple fix that applies a constant force along a vector dependent on the atoms' current orientation quaternion. I think this could have been achieved by using addforce and per-vector variables but adding this stuff to a fix seems like a cleaner solution and allows for more easy extension of more advanced features later.

## Author(s)

Stefan Paquay @ Brandeis University.

## Backward Compatibility

Yes; just adds new feature.

## Implementation Notes

This fix applies a constant force to an atom along the "body-frame" z-axis. Therefore, if the atom rotates (due to fix langevin for instance) the force also rotates with it.
One check I did is to make sure that, when using fix viscous with a constant friction, the velocity of non-interacting particles converges to the right velocity (given by force magnitude / friction coefficient).

## Post Submission Checklist

_Please check the fields below as they are completed_
- [X] The feature or features in this pull request is complete
- [X] Suitable new documentation files and/or updates to the existing docs are included
- [X] One or more example input decks are included
- [X] The source code follows the LAMMPS formatting guidelines

Pakketeretet2 commented:
@akohlmey 
The rename is nice, will adopt.

With regards to the second point, it seems like spherical particles do not have an orientation, only an angular frequency. To get an orientation in there a user would have to provide a quaternion or the fix itself should provide it. Can both be done but it is a lot of extra work, also because then the fix will have to take care of properly time-integrating the rotation quaternion.

For point particles the already-existing velocity vector could be used, that should be fairly straightforward to add. I have seen one publication that goes this route but most I know involve Brownian rotation. I figured I would just play safe and explicitly not allow things that do not have orientation, but I agree that that is a little strict.

akohlmey commented:
On Fri, Oct 26, 2018 at 4:01 PM Stefan Paquay <notifications@github.com>
wrote:

> @akohlmey <https://github.com/akohlmey>
> The rename is nice, will adopt.
>
> With regards to the second point, it seems like spherical particles do not
> have an orientation, only an angular frequency.
>
yes, i mixed something up with how things were handled a long time ago.
but atom style "body" also provides an orientation quaternion and could be
supported.

> To get an orientation in there a user would have to provide a quaternion
> or the fix itself should provide it. Can both be done but it is a lot of
> extra work, also because then the fix will have to take care of properly
> time-integrating the rotation quaternion.
>
this is overkill and i don't see a use case.

> For point particles the already-existing velocity vector could be used,
> that should be fairly straightforward to add. I have seen one publication
> that goes this route but most I know involve Brownian rotation.
>
then it would be a good idea to point this out in the documentation.

> I figured I would just play safe and explicitly not allow things that do
> not have orientation, but I agree that that is a little strict.
>
in order to avoid people making mistakes by accident. there should be a
(required) flag setting the mode. so that even ellipsoids/bodies could be
treated like point particles, if the corresponding flag is chosen. that
also leaves the door open for other "modes".

axel.




> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lammps/lammps/pull/1178#issuecomment-433526630>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARp41qWuFYp36gGUhlhJiUMlDiuWUziks5uo2n9gaJpZM4X6kfP>
> .
>


-- 
Dr. Axel Kohlmeyer  akohlmey@gmail.com  http://goo.gl/1wk0
College of Science & Technology, Temple University, Philadelphia PA, USA
International Centre for Theoretical Physics, Trieste. Italy.

sjplimp commented:
@akohlmey @Pakketeretet2 I agree with the name change suggestion, active is too generic.
I suggest:
a) add context to the doc page, i.e. what is this fix good for?  E.g. simple model of self-propelled particle motion.
b) to avoid user errors, I think the code should check if quats exist for the particles
(ellipsoids or body particles0.
Maybe a new atom->quat_exists() method should be added.  I don't see how in the
constructor you can cast avec to AtomVecEllipsoid*.  Won't that just give garbage if the
atom_style is not ellipsoid?
c) Also, atom_style ellipsoid does not require all particles be ellipsoids (and thus have a quat).
See how fix_nve_aspher.cpp checks in its constructor that you are not calling it
with a group that has non-ellipsoids.  Else you will access bonus[ellipsoid[i]] with a -1.

pdebuyl commented:
Hi @Pakketeretet2 does the fix handle two dimensional simulations?

akohlmey commented:
@Pakketeretet2 this has been sitting idle for a couple of months. Any updates/news/problems?

Pakketeretet2 commented:
Nothing new for a while, sorry. Last weeks have been busy with non-work
stuff (and the coming weeks will be busy too). :(

On Mon, Jan 14, 2019, 6:20 PM Axel Kohlmeyer <notifications@github.com
wrote:

> @Pakketeretet2 <https://github.com/Pakketeretet2> this has been sitting
> idle for a couple of months. Any updates/news/problems?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lammps/lammps/pull/1178#issuecomment-454201538>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQq0_HPG9aN3nzNvKpiblfnUTmFXAm6Eks5vDRCygaJpZM4X6kfP>
> .
>

akohlmey commented:
@Pakketeretet2 checking in again? any news? could we perhaps first add a minimal version of this and expand on it later?

Pakketeretet2 commented:
Nothing new yet except that I will actually have time to work on this next
week.

On Wed, Mar 20, 2019 at 10:18 AM Axel Kohlmeyer <notifications@github.com>
wrote:

> @Pakketeretet2 <https://github.com/Pakketeretet2> checking in again? any
> news? could we perhaps first add a minimal version of this and expand on it
> later?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lammps/lammps/pull/1178#issuecomment-474849854>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQq0_GMP8McxYoTngonruvhgRe55xE6sks5vYkMlgaJpZM4X6kfP>
> .
>

Pakketeretet2 commented:
I finally got some work done on this. It is not ready yet. The mode idea makes a lot of sense. I renamed it to propel/self, is that acceptable or are you more insistent on propel/atom?

sjplimp commented:
I like fix propel/self

jewettaij commented:
Cool.  Glad someone is working on this!

Incidentally, I am working on a [fix](https://github.com/jewettaij/lammps_mca_examples) (similar to fix bond/react) that allows atom types to change in response to collision events and other events.  (...For example, a message could be passed to an atom from a bonded neighbor causing it to change type.  You can make more complicated decisions by linking multiple atoms together.)  It would be cool if these two fixes work well together.

1) Feature request:  Would it be difficult to add an optional "type" (or "types") keyword that restricts the fix to only work on atoms of a given type (or range of types)?

2) I noticed that the "fix_propel_self.txt" file might need to be updated because it still refers to fix "active".
https://github.com/Pakketeretet2/lammps/blob/fix-active/doc/src/fix_propel_self.txt

Hmm so it seems "fix active" is available again.  Hmm...  (I probably won't use this name for my fix, but it's tempting.)

Pakketeretet2 commented:
> Cool. Glad someone is working on this!
> 
> Incidentally, I am working on a [fix](https://github.com/jewettaij/lammps_mca_examples) (similar to fix bond/react) that allows atom types to change in response to collision events and other events. (...For example, a message could be passed to an atom from a bonded neighbor causing it to change type. You can make more complicated decisions by linking multiple atoms together.) It would be cool if these two fixes work well together.
> 
>     1. Feature request:  Would it be difficult to add an optional "type" (or "types") keyword that restricts the fix to only work on atoms of a given type (or range of types)?
Hmm. Since you are asking I presume that's because particles will dynamically change type and this cannot be handled (efficiently) with a dynamical group? It can probably be done and should not be too hard. Please keep nagging, that way I will keep working on this. ;)

>     2. I noticed that the "fix_propel_self.txt" file might need to be updated because it still refers to fix "active".
>        https://github.com/Pakketeretet2/lammps/blob/fix-active/doc/src/fix_propel_self.txt

Thanks!

jewettaij commented:
> > Cool. Glad someone is working on this!
> > Incidentally, I am working on a [fix](https://github.com/jewettaij/lammps_mca_examples) (similar to fix bond/react) that allows atom types to change in response to collision events and other events. (...For example, a message could be passed to an atom from a bonded neighbor causing it to change type. You can make more complicated decisions by linking multiple atoms together.) It would be cool if these two fixes work well together.
> > ```
> > 1. Feature request:  Would it be difficult to add an optional "type" (or "types") keyword that restricts the fix to only work on atoms of a given type (or range of types)?
> > ```
> 
> Hmm. Since you are asking I presume that's because particles will dynamically change type and this cannot be handled (efficiently) with a dynamical group?

Yep.

> It can probably be done and should not be too hard. Please keep nagging, that way I will keep working on this. ;)

If you decide to implement this, the code in fix_atom_swap.cpp (between lines 165 and 174) shows a good way to enter atom types.  Here is the relevant excerpt:

```
void FixAtomSwap::options(int narg, char **arg) {
  :
  int iarg = 0;
  while (iarg < narg) {
    :
    } else if (strcmp(arg[iarg],"types") == 0) {
      if (iarg+3 > narg) error->all(FLERR,"Illegal fix atom/swap command");
      iarg++;
      while (iarg < narg) {
        if (isalpha(arg[iarg][0])) break;
        if (nswaptypes >= atom->ntypes) error->all(FLERR,"Illegal fix atom/swap command");
        type_list[nswaptypes] = force->numeric(FLERR,arg[iarg]);
        nswaptypes++;
        iarg++;
      }
    }
    : 
}
```
Alternatively, you can also use force->bounds(), to handle expressions like "4*7", but that restricts you to atom types which are consecutive:
```
force->bounds(FLERR, arg[iarg+1], atom->ntypes, atype_lo, atype_hi);
// If arg[iarg+1] == "4*7", then this would set atype_lo to 4 and atype_hi to 7
```

Cheers

> > ```
> > 2. I noticed that the "fix_propel_self.txt" file might need to be updated because it still refers to fix "active".
> >    https://github.com/Pakketeretet2/lammps/blob/fix-active/doc/src/fix_propel_self.txt
> > ```
> 
> Thanks!

pdebuyl commented:
Hey @Pakketeretet2 

I'd like to review the feature. What is the best way? In the comments here, in the commits, via email, the lammps mailing-list?

pdebuyl commented:
The example `active_viscous.in` does not seem to have noise for the orientation of the particles. "fix langevin" is not present in the file and that was what generated the rotational diffusion.

pdebuyl commented:
For reference, I believe that the 2d equivalent literature start at "Fily and Marchetti, Athermal Phase Separation of Self-Propelled Particles with No Alignment, PRL 108, 235702 (2012), http://dx.doi.org/10.1103/PhysRevLett.108.235702 ". I will check if there is a well identified 3D equivalent in the literature. Of course, these are stricly overdamped.

I have doubts regarding the use of the "non quaternion" version, where the velocity is used in the sense that it does not correspond to a widespread model. What references do you have in mind?

Pakketeretet2 commented:
> For reference, I believe that the 2d equivalent literature start at "Fily and Marchetti, Athermal Phase Separation of Self-Propelled Particles with No Alignment, PRL 108, 235702 (2012), http://dx.doi.org/10.1103/PhysRevLett.108.235702 ". I will check if there is a well identified 3D equivalent in the literature. Of course, these are stricly overdamped.
> 
> I have doubts regarding the use of the "non quaternion" version, where the velocity is used in the sense that it does not correspond to a widespread model. What references do you have in mind?

I remember seeing an article where they did that (I think by Mike Cates' group), but I cannot find it anymore.

Pakketeretet2 commented:
I added some more changes, I think code-wise it is now pretty much done (apart from the feature @jewettaij requested) but we probably should test it physics-wise. 

@pdebuyl For review, comments here are fine I think and probably the best way to do it because then it is truly open development.

Pakketeretet2 commented:
> The example `active_viscous.in` does not seem to have noise for the orientation of the particles. "fix langevin" is not present in the file and that was what generated the rotational diffusion.

Ignore that example, the latest moved to examples/USER/misc/fix_propel_self to keep in sync with the fix name change. For the velocity mode the changes in direction are now purely due to the external and internal forces.
For the 3D example with quaternions I added the angular momentum to the thermostat.

pdebuyl commented:
Is the article by the Cates group "Fodor et al, How Far from Equilibrium Is Active Matter?, PRL 117, 038103 (2016)" http://dx.doi.org/10.1103/PhysRevLett.117.038103

Pakketeretet2 commented:
> Is the article by the Cates group "Fodor et al, How Far from Equilibrium Is Active Matter?, PRL 117, 038103 (2016)" http://dx.doi.org/10.1103/PhysRevLett.117.038103

No, it is not that one. I asked Yaouen Fily (we shared an office for a bit at Brandeis), maybe he remembers if such a paper/model exists.

Pakketeretet2 commented:
Yaouen could not remember such a paper so maybe I mixed it up. He did mention some models where the active force attempts to align to the velocity over time, so the velocity mode can be considered a limiting case of that. This is now added to the documentation.
@jewettaij, I added the types option as well. You can just specify a list of types to which the fix should be applied. This is also shown in the in.2d_langevin example.
At this point, do you guys have good suggestions for tests of the implementation? Theoretical expressions for MSD or whatever that should hold?

jewettaij commented:
The recent discussion on the mailing list today gave me an idea.  I don't think it's hard to implement, but I wish I'd thought of it earlier:

Would it be difficult to apply a force to all pairs of atoms of different type which are connected by a particular type of bond?  The fix would apply a force along the direction of the bond axis.  (Using two different atom types to determine the polarity.)  This way you don't have to worry about additional properties to the atom (such as quaternions).  You could create a new keyword to enable this behavior ("bond").  The syntax might look something like this:

fix group propell_self bond bondType atomType1 atomType2

Either way, thanks again for working on this cool thing (useful fix).
-Andrew

pdebuyl commented:
@jewettaij do you have a reference to such a model? I still believe that as a first step LAMMPS should propose a known model.

The typical active brownian particle are defined by the self-propelled velocity and their orientational persistence. The dynamical characterization is given by the autocorrelation of the unit orientation vector. In order to check all of that, I dug into the code of fix langevin and fix nve asphere.

I can predict properly the angular momentum autocorrelations, so I guess that the friction is well given by "inertia * ascale / t_period" (those are the parameters to fix langevin). Using "D_r  = kB T / friction" for the autocorrelation of the unit orientation vector does not give the proper relaxation time. Did anyone tested this in LAMMPS already?

jewettaij commented:
Hi pdebuyl
I don't have a reference for you.  I just mentioned it because that approach might be useful to apply a propulsion force in an arbitrary direction to objects with arbitrary shapes instead of only ellipsoids.  (LAMMPS allows users to add bonds between ellipsoids, but to my knowledge, there isn't a way to prevent those ellipsoids from spinning freely inside the molecule yet, so users wouldn't be able to control the direction of the propulsion.  However I think the USER-MOLC authors are working on addressing this particular issue right now.)  Either way, I think having the ellipsoids supported by this fix a good idea too.

pdebuyl commented:
I am making progress on a "validation" of the fix.

Apart from assessing the code directly, my plan is to

1. Define suitable conditions to reproduce overdamped Brownian motion. One important feature here is that the rotational motion is characterized by <u(tau) u> = exp(-t / tau_r) where tau_r = gamma_r /2
2. Relate the physical quantities above to the fix langevin parameters damp kT and angmom.
3. Verify the MSD for a dilute (non-interacting) system.

Optionally, I could also try a replication of one of the ABP papers but I believe that the steps above should suffice to warrant inclusion in LAMMPS.

pdebuyl commented:
Hi @Pakketeretet2 

I am mostly there for the quaternion scheme, actually running checks.

The quaternion scheme rotates the force term from (0,0,1) to the actual force direction. With fix enforce2d, the quaternion will always define a rotation around the axis 1z however. I propose to change 
 
```.diff
-      double f_act[3] = { 0.0, 0.0, 1.0 };
+      double f_act[3] = { 1.0, 0.0, 0.0 };
```

I just ran a simulation that confirms the expected MSD with this change. (I'd be happy to submit a PR, let me know). It is important to know, of course, that this propels the particle in a different direction (x instead of z) which is important for non-spherical active particles. What do you think? (2D ABP are really the most important in the literature). An option would be to have a user-defined basic orientation for the active force.

pdebuyl commented:
I have performed the following tests, in 2D and 3D.

For the quaternion scheme, provided that the damp time is small enough, the rotational
autocorrelation is exponential: exp(-D_r t) in 2D and exp(-2 D_r t) in 3D.

The effective diffusion coefficient for self-propelled particles is thus D + v0**2 tau_r / d
where d is the dimension of the problem. I verified this numerically, as well as checking
the value for rotational friction and the exponential damping of the angular momenta.

For the velocity scheme, the velocity process can be mapped to an overdamped equation of
motion with effective friction m**2/gamma. |v| thus admits a stationary equilibrium
distribution for the effective potential m (v**2/2 + v0 |v|), in 2D and 3D.

I don't have a reference theory for the MSD of the velocity scheme however.

Remarks:
- the velocity scheme corresponds to a velocity dependent friction of gamma(v) = gamma_0 *
  (1 - v0/|v|), one of the forms mentioned in Erdmann et al, Eur. Phys. J. B 15, 105-113
  (2000). The authors suggest to settle instead for gamma_0 * (1 - q/(1+d v**2)). I am glad
  I found the same expression in the paper as in the code already.
- I propose a change for the alignment of the quaternion force.

I had at look at the source code itself. In addition to the change to the orientation, I would like to propose a "default" setup in the doc as well, explaining the relation to the ABP model.

pdebuyl commented:
So, I recommend merging :-)

pdebuyl commented:
@Pakketeretet2 I submitted a PR to your branch for the documentation already. Concerning the orientation along either x or z, I wait for your feedback.

Pakketeretet2 commented:
Thanks for the doc changes! This has been lying around for a while, sorry about that. For the proposed change with respect to z vs. x, sure, we can just change it to point in the body's "x"-direction initially. That should make it work out of the box in 2D as well. We'll just have to update the docs.

pdebuyl commented:
With pleasure :-) I have a new project that uses it!

I filed a PR for the orientation. The code is fine and I recommend to merge then.

For new features in MD codes, I am a supporter of writing a paper or a report on their use. Have you though of such work? There are a few papers using LAMMPS for active Brownian particles but I couldn't find any where all the details are provided or the source code of course! Besides, once this PR is merged, this feature would be part of the LAMMPS distribution, hence the need for a reference analysis.

pdebuyl commented:
Hi, just a reminder :-)

I have started to use the fix and still support merging the PR (after @Pakketeretet2 merge the PR at https://github.com/Pakketeretet2/lammps/pull/5 ).

pdebuyl commented:
If necessary, I am also willing to rebase against master to fix the conflicts in the doc files.

pdebuyl commented:
@Pakketeretet2 I rebased my branch at https://github.com/Pakketeretet2/lammps/pull/5 it should merge fine with lammps master

gokulup009 commented:
I am getting segmentation fault (core dumped) when using the fix propel/self how to fix this problem .please help me.

akohlmey commented:
> I am getting segmentation fault (core dumped) when using the fix propel/self how to fix this problem .please help me.

Please file a proper bug report in the Issues section with sufficient information about your environment and a (simple!) test case to reproduce it.
